### DISCUSSION TITLE
Log throttling, errors and enrichers

### DISCUSSION THREAD

**Discussion by Tthongsy** [datalust/seq-tickets#1305] [Help] [answered]
Hi, I was wondering if there was a way to setup log throttling and if there was a way to send emails out when the same type of log occurs more than a set amount of times?

I am also getting a few errors within the ingestion logs and I remember you told me there would be more information found in newer versions of Seq. However I still get the same type of information and I'm unsure where to find the more detailed information within the ingestion logs?

The errors:
**Rejecting payload from x.x.x.x due to invalid JSON, request body could not be parsed
System.IO.IOException
 ---> Microsoft.AspNetCore.Server.HttpSys.HttpSysException (1229): An operation was attempted on a nonexistent network connection.
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.HttpSys.RequestStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.IO.StreamReader.ReadBuffer(Span`1 userBuffer, Boolean& readToUserBuffer)
   at System.IO.StreamReader.ReadSpan(Span`1 buffer)
   at System.IO.StreamReader.Read(Char[] buffer, Int32 index, Int32 count)
   at Newtonsoft.Json.JsonTextReader.ReadData(Boolean append, Int32 charsRequired)
   at Newtonsoft.Json.JsonTextReader.ParseValue()
   at Newtonsoft.Json.JsonReader.ReadForType(JsonContract contract, Boolean hasConverter)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonSerializer.Deserialize[T](JsonReader reader)
   at Seq.Server.Web.Api.IngestionController.TryParseRawFormatBody(Input input, InputSettings inputSettings, StorageEventCreationData[]& events, String& errorMessage)**

For this error below, if I recall, you stated that this is something to not be worried about unless there is reason for low bandwidth or disk space?
**Exception raised when writing to event storage
Flare.FlareCapacityException: The stream is temporarily read-only.
   at Flare.Storage.StorageEngine.AcquireWriteLock()
   at Flare.Storage.StorageEngine.Add(StructuredEvent[] events)
   at Flare.Events.EventStore.Add(StorageEventCreationData[] eventsCreationData)
   at Flare.Queries.DataStore.Add(StorageEventCreationData[] eventsCreationData)
   at Seq.Server.Web.Api.IngestionController.Ingest()**

Thanks,
Tommy T

**Comment by nblumhardt** [comment, marked as answer]
Hi @Tthongsy,

> I was wondering if there was a way to setup log throttling

No, not really - filtering (or level control) is the only option in this area right now.

> there was a way to send emails out when the same type of log occurs more than a set amount of times?

An [alert](https://docs.datalust.co/docs/alerts) can do this; if you want to do this by event type for all events, then adding `@EventType` to the `group by` section of the corresponding dashboard chart, and `count > 100` (or whatever limit you want) as the alert condition, is the way to go.

> Rejecting payload from x.x.x.x due to invalid JSON, request body could not be parsed

There's probably not a lot more info that can be surfaced for this particular error; it shows up when a network connection is dropped for any reason, and is generally safe to ignore (it would only be of interest if you're having log delivery issues).

> Flare.FlareCapacityException: The stream is temporarily read-only.

Yes, this one is nothing to worry about if you see it occasionally; it's triggered when a large number of writes arrive at the same time, and prevents resource exhaustion on the server. If you see it more than a few times per day, faster disks are likely to be required.

Let me know if this helps!

Best regards,
Nick

**Reply by Tthongsy** [reply]
Is there a way to set up custom levels and to do alerts for these custom levels?

**Reply by nblumhardt** [reply]
@Tthongsy yes, any string in `@Level` should be treated as valid by Seq. E.g.:

```
@Level = 'Broken'
```

will work if your apps use "Broken" to mean "Error".

If you want the "Warnings" or "Errors" signals to recognize custom level names, you should be able to do that by editing them, and adding your custom level names to the filters associated with each.

![image](https://user-images.githubusercontent.com/342712/129987282-ce6f8c8d-2cdd-41d8-a515-2f58b1db34f2.png)

Let me know if this helps,

Best regards,
Nick

**Reply by Tthongsy** [reply]
Thanks,

Is there also any way with the above exceptions and errors to find the cause of it within the ingestion logs?

**Reply by nblumhardt** [reply]
Hi!

For the `FlareCapacityException`, the best way to get a picture of what's going on there is via the _Ingestion_ screen; you're likely to see some ingestion spikes around the times that requests are being rejected, and by examining the stream (or the per-API-key ingestion charts) at that point in time, see where the influx of data is coming from.

There isn't much else to find regarding the "nonexistent network connection" message - but it's also generally safe to ignore.

Are you chasing down some missing log events?

**Reply by Tthongsy** [reply]
Yea, I'm just looking to map where the source of these errors are coming from. I wasn't sure if there was any easier way to go about it?

**Reply by nblumhardt** [reply]
Thanks for your reply :+1:

The ingestion log's really only intended for diagnostics, e.g. when expected log events are missing, so we don't filter out errors from transient or benign conditions. It's very normal to see a few network-related log events in there - there's no need to take action on them unless it's part of troubleshooting a larger issue.

It's likely that every service logging to Seq will occasionally experience a connection drop-out, or contact the server during a spike of ingestion traffic/congestion.

If you're missing log events that you expect to find in Seq, then mapping out the approximate times of the missing events, and trying to correlate these with entries in the ingestion log or in the full diagnostic report is the way to go.

Hope this helps!
Nick
